### PR TITLE
Consolidate Execution Policies

### LINKED ISSUE #56: Consolidate mint::policy and spin::execution_space in axom::core
The ``spin::execution_space`` defines a set of high-level execution policies, e.g., ``spin::SEQ_EXEC``, ``spin::CUDA_EXEC``, etc. and a traits class that binds the high-level  execution policies to the underlying ``RAJA`` execution policies and to a corresponding ``Umpire`` memory space.

A similar strategy is employed by ``mint::policy`` for Mint's mesh-aware execution model.

Although there is overlap between ``mint::policy`` and ``spin::execution_space`` the two implementations also complement each other. For example, ``mint::policy_traits`` has mappings to corresponding ``RAJA`` kernel and synchronization policies that are not implemented by ``spin::execution_space``. Likewise, the ``spin::execution_space`` has mappings to  ``RAJA`` atomic policies and default memory space which are not implemented by``mint::policy_traits``.

As more Axom components, e.g., ``Quest``, begin to support fine-grain parallelism on heterogeneous architectures, using ``RAJA``'s parallel execution model and ``Umpire`` for memory management, a similar traits class and policies will most likely be needed.

This issue proposes to unify ``spin::execution_space`` and ``mint::policy`` in ``axom::core`` so that the net functionality and future extensions are broadly available to all Axom components.

### PR DESCRIPTION
**Pull request by gzagaris**
# Summary

Consolidates the execution policies in axom/core. See #56 for details.

Resolves #56

### CODE REVIEW

File: src/axom/core/execution/internal/cuda_exec.hpp [hunk 1]
```diff
@@ -0,0 +1,140 @@
...
+ * \tparam BLOCK_SIZE the number of CUDA threads to launch
+ *
+ */
+template < int BLOCK_SIZE >
+struct execution_space< CUDA_EXEC< BLOCK_SIZE, SYNCHRONOUS > >
+{
+  using loop_policy   = RAJA::cuda_exec< BLOCK_SIZE >;
+
+  /* *INDENT-OFF* */
+  using loop2d_policy =
```
Comment (rhornung67) [inline src/axom/core/execution/internal/cuda_exec.hpp]: Are these policy definitions too constraining, potentially? They might be suboptimal depending on the kernel.

File: src/axom/core/execution/internal/omp_exec.hpp [hunk 2]
```diff
@@ -0,0 +1,73 @@
...
+/*!
+ * \brief execution_space traits specialization for OMP_EXEC
+ */
+template < >
+struct execution_space< OMP_EXEC >
+{
+  using loop_policy = RAJA::omp_parallel_for_exec;
+
+  /* *INDENT-OFF* */
+  using loop2d_policy = RAJA::KernelPolicy<
```
Comment (rhornung67) [inline src/axom/core/execution/internal/omp_exec.hpp]: For most kernels, using OpenMP collapse on all loop levels performs poorer than applying OpenMP to a subset.

File: src/axom/core/tests/core_execution_space.cpp [hunk 3]
```diff
@@ -0,0 +1,309 @@
...
+//------------------------------------------------------------------------------
+#ifdef AXOM_USE_RAJA
+TEST( core_execution_space, check_seq_exec )
+{
+  check_valid< axom::SEQ_EXEC >( );
+
+  constexpr bool IS_ASYNC = false;
+
+  /* *INDENT-OFF* */
+  using raja_loop2d_policy =
```
Comment (rhornung67) [inline src/axom/core/tests/core_execution_space.cpp]: Why use RAJA policies here and not the axom ones you defined?

Comment (gzagaris) [reply src/axom/core/tests/core_execution_space.cpp]: This code tests the mapping of the `execution_space` traits class to the corresponding RAJA policies. So, the expected RAJA policies are hard-coded in this test. Makes sense?

File: src/axom/core/execution/internal/cuda_exec.hpp (hunk 1 above)
Comment (gzagaris) [reply src/axom/core/execution/internal/cuda_exec.hpp]: The loop 2d/3d policies are intended for generic i-j-k loops. Is there something better and generic that we should do in this case instead?

These type of policies should not be used for things like, loop over the DOFs of each element and process each DOF independently.  Indeed, that would be suboptimal in this case. My expectation is that for specific use cases, different policies will be defined at the application level. 

That said, exposing these loop policies here, may be misleading application developers to use them regardless.

Perhaps, we shouldn't define these policies here at all. We could leave it to the application, perhaps, by inheriting and extending the `execution_space` traits class here. Thoughts?

Comment (rhornung67) [reply src/axom/core/execution/internal/cuda_exec.hpp]: That last part seems better, I think.

Comment (gzagaris) [reply src/axom/core/execution/internal/cuda_exec.hpp]: Sounds good -- I'll remove the predefined kernel policies from here.

File: src/axom/mint/execution/internal/structured_exec.hpp [hunk 4]
```diff
@@ -0,0 +1,157 @@
...
+  /* *INDENT-OFF* */
+   using loop2d_policy = RAJA::KernelPolicy<
+       RAJA::statement::Collapse< RAJA::omp_parallel_collapse_exec,
+                                  RAJA::ArgList< 1,0 >,
+                                  RAJA::statement::Lambda< 0 > > >;
+
+   using loop3d_policy = RAJA::KernelPolicy<
+       RAJA::statement::Collapse< RAJA::omp_parallel_collapse_exec,
+                                  RAJA::ArgList< 2,1,0 >,
+                                  RAJA::statement::Lambda< 0 > > >;
```
Comment (gzagaris) [inline src/axom/mint/execution/internal/structured_exec.hpp]: Is there something better, yet general (i.e., without knowledge of the mesh dimensions), that we can do here?

Comment (gzagaris) [reply src/axom/mint/execution/internal/structured_exec.hpp]: For example, we could predefine some Tiling policies, and at runtime, we could choose if we can use that or fall-back to the collapse?

Comment (gzagaris) [reply src/axom/mint/execution/internal/structured_exec.hpp]: I changed this to parallelize with OpenMP in the outer-loop and used tiling for CUDA.

File: src/axom/mint/docs/sphinx/sections/execution_model.rst [hunk 5]
```diff
@@ -82,32 +82,31 @@ Execution Policy
...
+| **SEQ_EXEC**                     |    None.                       | Sequential execution on         |
+|                                  |                                | the CPU.                        |
++----------------------------------+--------------------------------+---------------------------------+
+| **OMP_EXEC**                     |  `RAJA`_ + OpenMP              | Parallel execution on           |
+|                                  |                                | the CPU using OpenMP.           |
++----------------------------------+--------------------------------+---------------------------------+
+| **CUDA_EXEC< NBLOCKS >**         |  `RAJA`_ + CUDA +              | Parallel execution on           |
+|                                  |  `Umpire`_ (memory management) | CUDA-enabled GPUs.              |
++----------------------------------+--------------------------------+---------------------------------+
+| **CUDA_EXEC< NBLOCKS, ASYNC >**  |  `RAJA`_ + CUDA                | Asynchronous parallel           |
```
[the remark this answers is not in the document]
Comment (gzagaris) [reply src/axom/mint/docs/sphinx/sections/execution_model.rst]: Yes -- good catch! I pushed a fix.

[the remark this answers is not in the document]
Comment (rhornung67) [reply src/axom/mint/docs/sphinx/sections/execution_model.rst]: @gzagaris if your CUDA_EXEC< NBLOCKS > dipatches to something using a RAJA::cuda_exec policy the template argument is wrong. It is # threads / block, not # blocks. Also, RAJA provides a reasonable default of 256 if not specified by a user. This should be noted as well.

[the remark this answers is not in the document]
Comment (gzagaris) [reply src/axom/mint/docs/sphinx/sections/execution_model.rst]: Thanks -- I renamed the template parameter to `BLOCKSIZE`, which is what we actually call it in code.

[the remark this answers is not in the document]
Comment (rhornung67) [reply src/axom/mint/docs/sphinx/sections/execution_model.rst]: 👍

**Pull request merged**